File: configuration-manager-go-client/configuration_service.go
package configmanager

import (
	"context"
	"fmt"
	"net/url"
	"strconv"
)

// GetAll retrieves all configurations with optional filtering
func (c *ConfigurationService) GetAll(ctx context.Context, options *GetConfigurationsOptions) (*ConfigurationListResponse, error) {
	params := url.Values{}

	if options != nil {
		if options.Type != nil {
			params.Set("type", string(*options.Type))
		}
		if options.Status != nil {
			params.Set("status", string(*options.Status))
		}
		if options.IncludeArchived != nil {
			params.Set("includeArchived", strconv.FormatBool(*options.IncludeArchived))
		}
	}

	resp, err := c.client.GetWithQuery(ctx, "configs", params)
	if err != nil {
		return nil, fmt.Errorf("get configurations request failed: %w", err)
	}
	defer c.client.CloseResponse(resp)

	var response ConfigurationListResponse
	if err := c.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode configurations response: %w", err)
	}

	return &response, nil
}

// Get retrieves a specific configuration by ID
func (c *ConfigurationService) Get(ctx context.Context, id string, options *GetConfigurationOptions) (*ResolvedConfigurationResponse, error) {
	if id == "" {
		return nil, NewValidationError("configuration ID cannot be empty", 0)
	}

	params := url.Values{}
	if options != nil {
		if options.Provenance != nil && *options.Provenance {
			params.Set("provenance", "true")
		}
		if options.Raw != nil && *options.Raw {
			params.Set("raw", "true")
		}
	}

	path := fmt.Sprintf("configs/%s", id)
	resp, err := c.client.GetWithQuery(ctx, path, params)
	if err != nil {
		return nil, fmt.Errorf("get configuration request failed: %w", err)
	}
	defer c.client.CloseResponse(resp)

	var response ResolvedConfigurationResponse
	if err := c.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode configuration response: %w", err)
	}

	return &response, nil
}

// Create creates a new configuration
func (c *ConfigurationService) Create(ctx context.Context, request *CreateConfigurationRequest) (*ConfigurationResponse, error) {
	if request == nil {
		return nil, NewValidationError("create request cannot be nil", 0)
	}

	if request.Name == "" {
		return nil, NewValidationError("configuration name cannot be empty", 0)
	}

	resp, err := c.client.Post(ctx, "configs", request)
	if err != nil {
		return nil, fmt.Errorf("create configuration request failed: %w", err)
	}
	defer c.client.CloseResponse(resp)

	var response ConfigurationResponse
	if err := c.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode create configuration response: %w", err)
	}

	return &response, nil
}

// Update updates an existing configuration
func (c *ConfigurationService) Update(ctx context.Context, id string, request *UpdateConfigurationRequest) (*ConfigurationResponse, error) {
	if id == "" {
		return nil, NewValidationError("configuration ID cannot be empty", 0)
	}

	if request == nil {
		return nil, NewValidationError("update request cannot be nil", 0)
	}

	path := fmt.Sprintf("configs/%s", id)
	resp, err := c.client.Put(ctx, path, request)
	if err != nil {
		return nil, fmt.Errorf("update configuration request failed: %w", err)
	}
	defer c.client.CloseResponse(resp)

	var response ConfigurationResponse
	if err := c.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode update configuration response: %w", err)
	}

	return &response, nil
}

// Delete deletes a configuration
func (c *ConfigurationService) Delete(ctx context.Context, id string) (*ConfigurationResponse, error) {
	if id == "" {
		return nil, NewValidationError("configuration ID cannot be empty", 0)
	}

	path := fmt.Sprintf("configs/%s", id)
	resp, err := c.client.Delete(ctx, path)
	if err != nil {
		return nil, fmt.Errorf("delete configuration request failed: %w", err)
	}
	defer c.client.CloseResponse(resp)

	var response ConfigurationResponse
	if err := c.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode delete configuration response: %w", err)
	}

	return &response, nil
}

// GetValue gets configuration data at a specific path
func (c *ConfigurationService) GetValue(ctx context.Context, id string, options *GetConfigurationValueOptions) (*ConfigurationValueResponse, error) {
	if id == "" {
		return nil, NewValidationError("configuration ID cannot be empty", 0)
	}

	params := url.Values{}
	if options != nil {
		if options.Path != nil {
			params.Set("path", *options.Path)
		}
		if options.Minimal != nil && *options.Minimal {
			params.Set("minimal", "true")
		}
	}

	path := fmt.Sprintf("configs/%s/data", id)
	resp, err := c.client.GetWithQuery(ctx, path, params)
	if err != nil {
		return nil, fmt.Errorf("get configuration value request failed: %w", err)
	}
	defer c.client.CloseResponse(resp)

	var response ConfigurationValueResponse
	if err := c.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode configuration value response: %w", err)
	}

	return &response, nil
}

// GetValueByName gets configuration data at a specific path by configuration name
func (c *ConfigurationService) GetValueByName(ctx context.Context, configurationName string, options *GetConfigurationValueOptions) (*ConfigurationValueResponse, error) {
	if configurationName == "" {
		return nil, NewValidationError("configuration name cannot be empty", 0)
	}

	params := url.Values{}
	if options != nil {
		if options.Path != nil {
			params.Set("path", *options.Path)
		}
		if options.Minimal != nil && *options.Minimal {
			params.Set("minimal", "true")
		}
	}

	path := fmt.Sprintf("configs/by-name/%s/data", url.PathEscape(configurationName))
	resp, err := c.client.GetWithQuery(ctx, path, params)
	if err != nil {
		return nil, fmt.Errorf("get configuration value by name request failed: %w", err)
	}
	defer c.client.CloseResponse(resp)

	var response ConfigurationValueResponse
	if err := c.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode configuration value by name response: %w", err)
	}

	return &response, nil
}

// Commit commits a draft configuration
func (c *ConfigurationService) Commit(ctx context.Context, id string) (*ConfigurationResponse, error) {
	if id == "" {
		return nil, NewValidationError("configuration ID cannot be empty", 0)
	}

	path := fmt.Sprintf("configs/%s/commit", id)
	resp, err := c.client.Post(ctx, path, nil)
	if err != nil {
		return nil, fmt.Errorf("commit configuration request failed: %w", err)
	}
	defer c.client.CloseResponse(resp)

	var response ConfigurationResponse
	if err := c.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode commit configuration response: %w", err)
	}

	return &response, nil
}

// GetChildren gets child configurations
func (c *ConfigurationService) GetChildren(ctx context.Context, id string, options *GetChildConfigurationsOptions) (*ChildrenConfigurationResponse, error) {
	if id == "" {
		return nil, NewValidationError("configuration ID cannot be empty", 0)
	}

	params := url.Values{}
	if options != nil && options.IncludeArchived != nil && *options.IncludeArchived {
		params.Set("includeArchived", "true")
	}

	path := fmt.Sprintf("configs/%s/children", id)
	resp, err := c.client.GetWithQuery(ctx, path, params)
	if err != nil {
		return nil, fmt.Errorf("get child configurations request failed: %w", err)
	}
	defer c.client.CloseResponse(resp)

	var response ChildrenConfigurationResponse
	if err := c.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode child configurations response: %w", err)
	}

	return &response, nil
}

// GetComponents gets all components with their versions
func (c *ConfigurationService) GetComponents(ctx context.Context) (*ComponentsResponse, error) {
	resp, err := c.client.Get(ctx, "configs/components")
	if err != nil {
		return nil, fmt.Errorf("get components request failed: %w", err)
	}
	defer c.client.CloseResponse(resp)

	var response ComponentsResponse
	if err := c.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode components response: %w", err)
	}

	return &response, nil
}

// Rename renames a configuration (admin only)
func (c *ConfigurationService) Rename(ctx context.Context, id, newName string) (*ConfigurationResponse, error) {
	if id == "" {
		return nil, NewValidationError("configuration ID cannot be empty", 0)
	}

	if newName == "" {
		return nil, NewValidationError("new name cannot be empty", 0)
	}

	request := &RenameConfigurationRequest{Name: newName}

	path := fmt.Sprintf("configs/%s/rename", id)
	resp, err := c.client.Put(ctx, path, request)
	if err != nil {
		return nil, fmt.Errorf("rename configuration request failed: %w", err)
	}
	defer c.client.CloseResponse(resp)

	var response ConfigurationResponse
	if err := c.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode rename configuration response: %w", err)
	}

	return &response, nil
}

// Archive archives a configuration (admin only)
func (c *ConfigurationService) Archive(ctx context.Context, id string, archiveChildren *bool) (*ConfigurationResponse, error) {
	if id == "" {
		return nil, NewValidationError("configuration ID cannot be empty", 0)
	}

	request := &ArchiveConfigurationRequest{}
	if archiveChildren != nil {
		request.ArchiveChildren = archiveChildren
	}

	path := fmt.Sprintf("configs/%s/archive", id)
	resp, err := c.client.Post(ctx, path, request)
	if err != nil {
		return nil, fmt.Errorf("archive configuration request failed: %w", err)
	}
	defer c.client.CloseResponse(resp)

	var response ConfigurationResponse
	if err := c.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode archive configuration response: %w", err)
	}

	return &response, nil
}

// Restore restores an archived configuration (admin only)
func (c *ConfigurationService) Restore(ctx context.Context, id string) (*ConfigurationResponse, error) {
	if id == "" {
		return nil, NewValidationError("configuration ID cannot be empty", 0)
	}

	path := fmt.Sprintf("configs/%s/restore", id)
	resp, err := c.client.Post(ctx, path, nil)
	if err != nil {
		return nil, fmt.Errorf("restore configuration request failed: %w", err)
	}
	defer c.client.CloseResponse(resp)

	var response ConfigurationResponse
	if err := c.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode restore configuration response: %w", err)
	}

	return &response, nil
}

File: configuration-manager-python-client/configuration_manager_client/services/configuration_service.py
"""Configuration management service for Configuration Manager API."""

from typing import Optional
from urllib.parse import quote

from ..http_client import HTTPClient
from ..models.configuration import (
    GetConfigurationsOptions,
    ConfigurationListResponse,
    CreateConfigurationRequest,
    UpdateConfigurationRequest,
    ConfigurationResponse,
    ResolvedConfigurationResponse,
    GetConfigurationOptions,
    ConfigurationValueResponse,
    GetConfigurationValueOptions,
    ChildrenConfigurationResponse,
    GetChildConfigurationsOptions,
    ComponentsResponse,
    RenameConfigurationRequest,
    ArchiveConfigurationRequest,
)
from ..exceptions import ValidationError
from ..utils import from_dict, build_query_params


class ConfigurationService:
    """Configuration management service client."""

    def __init__(self, http_client: HTTPClient):
        """Initialize configuration service.
        
        Args:
            http_client: HTTP client instance
        """
        self.client = http_client

    def get_all(
        self, 
        options: Optional[GetConfigurationsOptions] = None
    ) -> ConfigurationListResponse:
        """Get all configurations with optional filtering.
        
        Args:
            options: Filtering options
            
        Returns:
            ConfigurationListResponse containing list of configurations
        """
        params = {}
        if options:
            if options.type:
                params["type"] = options.type.value
            if options.status:
                params["status"] = options.status.value
            if options.include_archived is not None:
                params["includeArchived"] = "true" if options.include_archived else "false"

        response = self.client.get("/configs", params=params)
        return from_dict(ConfigurationListResponse, response.json())

    def get(
        self, 
        config_id: str, 
        options: Optional[GetConfigurationOptions] = None
    ) -> ResolvedConfigurationResponse:
        """Get a specific configuration by ID.
        
        Args:
            config_id: Configuration ID
            options: Options for getting configuration
            
        Returns:
            ResolvedConfigurationResponse containing configuration data
            
        Raises:
            ValidationError: If config_id is empty
        """
        if not config_id:
            raise ValidationError("Configuration ID cannot be empty")

        params = {}
        if options:
            if options.provenance:
                params["provenance"] = "true"
            if options.raw:
                params["raw"] = "true"

        response = self.client.get(f"/configs/{config_id}", params=params)
        return from_dict(ResolvedConfigurationResponse, response.json())

    def create(self, request: CreateConfigurationRequest) -> ConfigurationResponse:
        """Create a new configuration.
        
        Args:
            request: Create configuration request
            
        Returns:
            ConfigurationResponse containing created configuration
            
        Raises:
            ValidationError: If request is invalid
        """
        if not request.name:
            raise ValidationError("Configuration name cannot be empty")

        data = {
            "name": request.name,
            "type": request.type.value,
            "data": request.data,
        }
        
        if request.parent_id:
            data["parent_id"] = request.parent_id
            
        if request.description:
            data["description"] = request.description

        response = self.client.post("/configs", json_data=data)
        return from_dict(ConfigurationResponse, response.json())

    def update(
        self, 
        config_id: str, 
        request: UpdateConfigurationRequest
    ) -> ConfigurationResponse:
        """Update an existing configuration.
        
        Args:
            config_id: Configuration ID
            request: Update configuration request
            
        Returns:
            ConfigurationResponse containing updated configuration
            
        Raises:
            ValidationError: If config_id is empty
        """
        if not config_id:
            raise ValidationError("Configuration ID cannot be empty")

        data = {}
        if request.data is not None:
            data["data"] = request.data
        if request.description is not None:
            data["description"] = request.description

        response = self.client.put(f"/configs/{config_id}", json_data=data)
        return from_dict(ConfigurationResponse, response.json())

    def delete(self, config_id: str) -> ConfigurationResponse:
        """Delete a configuration.
        
        Args:
            config_id: Configuration ID
            
        Returns:
            ConfigurationResponse containing deletion confirmation
            
        Raises:
            ValidationError: If config_id is empty
        """
        if not config_id:
            raise ValidationError("Configuration ID cannot be empty")

        response = self.client.delete(f"/configs/{config_id}")
        return from_dict(ConfigurationResponse, response.json())

    def get_value(
        self, 
        config_id: str, 
        options: Optional[GetConfigurationValueOptions] = None
    ) -> ConfigurationValueResponse:
        """Get configuration data at a specific path.
        
        Args:
            config_id: Configuration ID
            options: Options for getting configuration value
            
        Returns:
            ConfigurationValueResponse containing value data
            
        Raises:
            ValidationError: If config_id is empty
        """
        if not config_id:
            raise ValidationError("Configuration ID cannot be empty")

        params = {}
        if options:
            if options.path:
                params["path"] = options.path
            if options.minimal:
                params["minimal"] = "true"

        response = self.client.get(f"/configs/{config_id}/data", params=params)
        return from_dict(ConfigurationValueResponse, response.json())

    def commit(self, config_id: str) -> ConfigurationResponse:
        """Commit a draft configuration.
        
        Args:
            config_id: Configuration ID
            
        Returns:
            ConfigurationResponse containing commit confirmation
            
        Raises:
            ValidationError: If config_id is empty
        """
        if not config_id:
            raise ValidationError("Configuration ID cannot be empty")

        response = self.client.post(f"/configs/{config_id}/commit")
        return from_dict(ConfigurationResponse, response.json())

    def get_children(
        self, 
        config_id: str, 
        options: Optional[GetChildConfigurationsOptions] = None
    ) -> ChildrenConfigurationResponse:
        """Get child configurations.
        
        Args:
            config_id: Configuration ID
            options: Options for getting child configurations
            
        Returns:
            ChildrenConfigurationResponse containing child configurations
            
        Raises:
            ValidationError: If config_id is empty
        """
        if not config_id:
            raise ValidationError("Configuration ID cannot be empty")

        params = {}
        if options and options.include_archived:
            params["includeArchived"] = "true"

        response = self.client.get(f"/configs/{config_id}/children", params=params)
        return from_dict(ChildrenConfigurationResponse, response.json())

    def get_components(self) -> ComponentsResponse:
        """Get all components with their versions.
        
        Returns:
            ComponentsResponse containing components and versions
        """
        response = self.client.get("/configs/components")
        return from_dict(ComponentsResponse, response.json())

    def rename(self, config_id: str, new_name: str) -> ConfigurationResponse:
        """Rename a configuration (admin only).
        
        Args:
            config_id: Configuration ID
            new_name: New configuration name
            
        Returns:
            ConfigurationResponse containing renamed configuration
            
        Raises:
            ValidationError: If config_id or new_name is empty
        """
        if not config_id:
            raise ValidationError("Configuration ID cannot be empty")

        if not new_name:
            raise ValidationError("New name cannot be empty")

        data = {"name": new_name}
        response = self.client.put(f"/configs/{config_id}/rename", json_data=data)
        return from_dict(ConfigurationResponse, response.json())

    def archive(
        self, 
        config_id: str, 
        archive_children: Optional[bool] = None
    ) -> ConfigurationResponse:
        """Archive a configuration (admin only).
        
        Args:
            config_id: Configuration ID
            archive_children: Whether to archive child configurations
            
        Returns:
            ConfigurationResponse containing archive confirmation
            
        Raises:
            ValidationError: If config_id is empty
        """
        if not config_id:
            raise ValidationError("Configuration ID cannot be empty")

        data = {}
        if archive_children is not None:
            data["archiveChildren"] = archive_children

        response = self.client.post(f"/configs/{config_id}/archive", json_data=data)
        return from_dict(ConfigurationResponse, response.json())

    def restore(self, config_id: str) -> ConfigurationResponse:
        """Restore an archived configuration (admin only).
        
        Args:
            config_id: Configuration ID
            
        Returns:
            ConfigurationResponse containing restore confirmation
            
        Raises:
            ValidationError: If config_id is empty
        """
        if not config_id:
            raise ValidationError("Configuration ID cannot be empty")

        response = self.client.post(f"/configs/{config_id}/restore")
        return from_dict(ConfigurationResponse, response.json())
